backend/src/ee/services/audit-log/audit-log-queue.ts

```diff
@@ -42,88 +51,144 @@ export const auditLogQueueServiceFactory = ({
+              // request timeout
+              timeout: AUDIT_LOG_STREAM_TIMEOUT,
+              // connection timeout
+              signal: AbortSignal.timeout(AUDIT_LOG_STREAM_TIMEOUT)
+            });
+          }
+        )
+      );
     }
 
-    const plan = await licenseService.getPlan(orgId);
-    if (plan.auditLogsRetentionDays === 0) {
-      // skip inserting if audit log retention is 0 meaning its not supported
-      return;
+    // this is when audit log needs to write to db
+    if (job.name === QueueJobs.AuditLogWriteToDB) {
+      const lock = await keyStore.acquireLock([KeyStorePrefixes.AuditLogBatchLock], 5000, {
+        retryCount: 10,
+        retryDelay: 200
+      });
+      try {
+        const auditLogsInStore = await keyStore.listHeadPop(KeyStorePrefixes.AuditLogBatch, AUDIT_LOG_BATCH_SIZE);
+        if (!auditLogsInStore?.length || !Array.isArray(auditLogsInStore)) return;
+
+        const parsedAuditLogs = await AuditLogsSchema.omit({ id: true })
+          .array()
+          .parseAsync(
+            auditLogsInStore.map(
+              (el) =>
+                JSON.parse(el, (key, value) => {
+                  if (["updatedAt", "createdAt", "expiresAt"].includes(key)) return new Date(value as string);
+                  return value as unknown;
+                }) as unknown
+            )
+          );
+        await auditLogDAL.batchInsert(parsedAuditLogs);
+        const keyStoreAuditLogBatchSize = await keyStore.listLength(KeyStorePrefixes.AuditLogBatch);
+        // if the size is still not under AUDIT_LOG_BATCH_SIZE / 2 keep doing the compaction
+        if (keyStoreAuditLogBatchSize > AUDIT_LOG_BATCH_SIZE / 2) {
+          await queueService.queue(QueueName.AuditLog, QueueJobs.AuditLogWriteToDB, undefined, {
+            removeOnFail: true,
+            removeOnComplete: true,
+            delay: 50
+          });
+        }
+      } finally {
+        await lock.release();
+      }
     }
+  });
 
-    // For project actions, set TTL to project-level audit log retention config
-    // This condition ensures that the plan's audit log retention days cannot be bypassed
-    const ttlInDays =
-      project?.auditLogsRetentionDays && project.auditLogsRetentionDays < plan.auditLogsRetentionDays
-        ? project.auditLogsRetentionDays
-        : plan.auditLogsRetentionDays;
-
-    const ttl = ttlInDays * MS_IN_DAY;
-
-    const auditLog = await auditLogDAL.create({
-      actor: actor.type,
-      actorMetadata: actor.metadata,
-      userAgent,
-      projectId,
-      ipAddress,
-      orgId,
-      eventType: event.type,
-      expiresAt: new Date(Date.now() + ttl),
-      eventMetadata: event.metadata,
-      userAgentType
+  const startPeriodicAuditLogBatchWriteToDB = () =>
+    queueService.queue(QueueName.AuditLog, QueueJobs.AuditLogWriteToDB, undefined, {
+      removeOnFail: true,
+      removeOnComplete: true,
+      repeat: {
+        every: AUDIT_LOG_BATCH_WRITE_TO_DB_CYCLE_SPEED,
```

Comment:
So this will check for logs over filling every 5 seconds but there could be instances where the the RPS could be 1k per second. This would mean that by the time we try to do a check, there will already be 5k items or more under the redis key. 

Instead of doing a loop that keeps on running every 5 seconds (also could be wasted loops), we should trigger the batch write job after determining that we have met a threshold. What do you think?

Reply:
In those scenerio this will kick in 

https://github.com/Infisical/infisical/pull/2403/files/c7baec86841b17d7f1dba11a36e8b1a06c88e18c#diff-c2c2f65a2bc66e55f4b6f1623a270941d71e1d0f8fc58664d806c6010b3d27a5R163

That is there to ensure its below a threshold. 

Regarding waster cycles because it needed. We don't do batching until we reach a threshold. But what if the threshold is not met anytime soon. There could be documents that needs to inserted thus the periodic cycle kicks in.

Reply:
I see, so at 500 logs, it will create a job to flush. Got it

Reply:
Correction: 5K actually. I took average size of our audit log row and it was around 890 bytes.